File: сross_platform_software_systems/14/user_service/user_service/src/main/java/com/c3equalz/user_service/domain/user/services/AccessService.java
package com.c3equalz.user_service.domain.user.services;

import com.c3equalz.user_service.domain.common.services.BaseDomainService;
import com.c3equalz.user_service.domain.user.entities.User;
import com.c3equalz.user_service.domain.user.errors.ActivationChangeNotPermittedError;
import com.c3equalz.user_service.domain.user.errors.AuthorizationError;
import com.c3equalz.user_service.domain.user.errors.RoleChangeNotPermittedError;
import com.c3equalz.user_service.domain.user.events.UserChangedRoleEvent;
import com.c3equalz.user_service.domain.user.events.UserToggleActivationEvent;
import com.c3equalz.user_service.domain.user.services.authorization.Permission;
import com.c3equalz.user_service.domain.user.services.authorization.PermissionContext;
import com.c3equalz.user_service.domain.user.values.UserRole;

/**
 * Service for managing user access, roles, and authorization.
 */
public class AccessService extends BaseDomainService {

    /**
     * Toggles the admin role of a user.
     *
     * @param user the user to modify
     * @param isAdmin true to assign ADMIN role, false to assign USER role
     * @throws RoleChangeNotPermittedError if the user's role cannot be changed
     */
    public void toggleUserAdminRole(User user, boolean isAdmin) {
        if (!user.getRole().isChangeable()) {
            String msg = String.format(
                    "Changing role of user '%s' (%s) is not permitted.",
                    user.getName().getValue(),
                    user.getRole()
            );
            throw new RoleChangeNotPermittedError(msg);
        }

        UserRole oldRole = user.getRole();
        user.setRole(isAdmin ? UserRole.ADMIN : UserRole.USER);

        UserChangedRoleEvent newEvent = new UserChangedRoleEvent(
                user.getId(),
                oldRole,
                user.getRole()
        );

        recordEvent(newEvent);
    }

    /**
     * Toggles the activation status of a user.
     *
     * @param user the user to modify
     * @param isActive true to activate, false to deactivate
     * @throws ActivationChangeNotPermittedError if the user's activation cannot be changed
     */
    public void toggleUserActivation(User user, boolean isActive) {
        if (!user.getRole().isChangeable()) {
            String msg = String.format(
                    "Changing activation of user '%s' (%s) is not permitted.",
                    user.getName().getValue(),
                    user.getRole()
            );
            throw new ActivationChangeNotPermittedError(msg);
        }

        user.setActive(isActive);

        UserToggleActivationEvent newEvent = new UserToggleActivationEvent(
                user.getId(),
                isActive
        );

        recordEvent(newEvent);
    }

    /**
     * Authorizes a permission against a context.
     *
     * @param permission the permission to check
     * @param context the permission context
     * @param <PC> the type of permission context
     * @throws AuthorizationError if the permission is not satisfied
     */
    public <PC extends PermissionContext> void authorize(Permission<PC> permission, PC context) {
        if (!permission.isSatisfiedBy(context)) {
            throw new AuthorizationError("Not authorized.");
        }
    }
}

File: сross_platform_software_systems/14/chat_service/src/chat_service/domain/user/services/access_service.py
from chat_service.domain.common.services.base import DomainService
from chat_service.domain.user.entities.user import User
from chat_service.domain.user.errors import (
    ActivationChangeNotPermittedError,
    AuthorizationError,
    RoleChangeNotPermittedError,
)
from chat_service.domain.user.events import UserChangedRoleEvent, UserToggleActivationEvent
from chat_service.domain.user.services.authorization.base import (
    Permission,
    PermissionContext,
)
from chat_service.domain.user.services.authorization.constants import AUTHZ_NOT_AUTHORIZED
from chat_service.domain.user.values.user_role import UserRole


class AccessService(DomainService):
    def __init__(self) -> None:
        super().__init__()

    def toggle_user_admin_role(self, user: User, *, is_admin: bool) -> None:
        """
        :raises RoleChangeNotPermitted:
        """
        if not user.role.is_changeable:
            msg: str = f"Changing role of user {user.name.value!r} ({user.role}) is not permitted."
            raise RoleChangeNotPermittedError(msg)

        old_role: str = user.role

        user.role = UserRole.ADMIN if is_admin else UserRole.USER

        new_event: UserChangedRoleEvent = UserChangedRoleEvent(
            user_id=user.id,
            old_role=old_role,
            new_role=user.role,
        )

        self._record_event(new_event)

    def toggle_user_activation(self, user: User, *, is_active: bool) -> None:
        """
        :raises ActivationChangeNotPermitted:
        """
        if not user.role.is_changeable:
            msg: str = f"Changing activation of user {user.name.value!r} ({user.role}) is not permitted."
            raise ActivationChangeNotPermittedError(msg)

        user.is_active = is_active

        new_event: UserToggleActivationEvent = UserToggleActivationEvent(
            user_id=user.id,
            is_active=is_active,
        )

        self._record_event(new_event)

    # noinspection PyMethodMayBeStatic
    def authorize[PC: PermissionContext](
        self,
        permission: Permission[PC],
        *,
        context: PC,
    ) -> None:
        """
        :raises AuthorizationError:
        """
        if not permission.is_satisfied_by(context):
            raise AuthorizationError(AUTHZ_NOT_AUTHORIZED)